### PR TITLE
Ensure realize_frame does not overwrite representation class.

### PR DESCRIPTION
**Pull request by mhvk**
fixes #6208

And also an alternative to #6208 which I think is slightly cleaner.

### CODE REVIEW

**Comment by astrofrog** [comment]
Did you mean to reference https://github.com/astropy/astropy/pull/6208?

**Comment by mhvk** [comment]
Yes, now updated the top entry so that we close the proper one if this is merged!

**Comment by mhvk** [comment]
@Cadair - apart from review, I of course mostly hope to hear that this in fact solves the regression!

File: astropy/coordinates/baseframe.py [hunk 1]
```diff
@@ -30,11 +30,15 @@
                              SphericalRepresentation,
                              UnitSphericalRepresentation,
                              REPRESENTATION_CLASSES)
+from .frame_attributes import FrameAttribute
 
-# Import all FrameAttributes so we don't break backwards-compatibility
+# Import all orher FrameAttributes so we don't break backwards-compatibility
```
Comment (adrn) [inline astropy/coordinates/baseframe.py]: Typo: other

File: astropy/coordinates/baseframe.py [hunk 2]
```diff
@@ -30,11 +30,15 @@
...
 # (some users rely on them being here, although that is not
 # encouraged, as this is not the public API location.)
-from .frame_attributes import *
+from .frame_attributes import (
```
Comment (adrn) [inline astropy/coordinates/baseframe.py]: Why did you change this back? The argument for * was that we then don't have to remember to update this if we add any more `FrameAttribute` classes...

File: astropy/coordinates/baseframe.py [hunk 3]
```diff
@@ -530,7 +535,13 @@ def realize_frame(self, representation):
             A new object with the same frame attributes as this one, but
             with the ``representation`` as the data.
         """
-        return self._apply('replicate', _framedata=representation)
+        # Here we pass representation=None to _apply, since we do not want
+        # to insist that the realized frame has the same representation as
+        # self.  [Avoids breaking sunpy; see gh-6208]
+        # TODO: expose this? But unhandy change of name, since in the
```
Comment (adrn) [inline astropy/coordinates/baseframe.py]: In #6219 I added `representation_cls` and am planning to deprecate `representation` - I guess you could use `_representation_cls` just to be consistent? I don't really care.

File: astropy/coordinates/baseframe.py [hunk 4]
```diff
@@ -823,6 +834,8 @@ def _apply(self, method, *args, **kwargs):
         else:
             data = self.data if self.has_data else None
 
+        representation = kwargs.pop('_repr_cls', self.representation)
```
Comment (adrn) [inline astropy/coordinates/baseframe.py]: See comment above.

File: astropy/coordinates/baseframe.py (hunk 1 above)
Comment (mhvk) [inline astropy/coordinates/baseframe.py]: Done

File: astropy/coordinates/baseframe.py (hunk 2 above)
Comment (mhvk) [inline astropy/coordinates/baseframe.py]: This was partially to keep my is-something-imported checker happy, but mostly because ideally people do *not* import the attributes from `baseframe`. This way, we don't break backwards compatibility, but if we define a new attribute and someone wants to use it, it will break, and they'll realize that they should change...

File: astropy/coordinates/baseframe.py (hunk 3 above)
Comment (mhvk) [inline astropy/coordinates/baseframe.py]: Ah, that would be great! It is such a pain as is. So, I changed the name here, but didn't yet expose it, as I'd like to focus on just fixing the regression.

File: astropy/coordinates/baseframe.py (hunk 4 above)
Comment (mhvk) [inline astropy/coordinates/baseframe.py]: Done.

File: astropy/coordinates/tests/test_regression.py [hunk 5]
```diff
@@ -360,10 +362,68 @@ def test_regression_5743():
...
+        default_representation = CartesianRepresentation
+        my_attr = QuantityFrameAttribute(default=0, unit=u.m)
+
+    class MySpecialFrame(MyFrame):
+        def __init__(self, *args, **kwargs):
+            _rep_kwarg = kwargs.get('representation', None)
+            super(MyFrame, self).__init__(*args, **kwargs)
+            if not _rep_kwarg:
+                self.representation = self.default_representation
+                self._data = self.data.represent_as(self.representation)
```
Comment (Cadair) [inline astropy/coordinates/tests/test_regression.py]: So we don't actually call `represent_as` in the `__init__` we just set the `representation` attribute and set `_data` equal to the rep we make.

Comment (mhvk) [inline astropy/coordinates/tests/test_regression.py]: Happy to hear this fixes the problem!  Note that this is the `BaseRepresentation.represent_as` which should be equivalent to the explicit initialization you do in `sunpy` -- I just wanted to make an example that didn't depend on it.

Comment (Cadair) [inline astropy/coordinates/tests/test_regression.py]: ah cool ok.
